torch/csrc/jit/codegen/cuda/glfdc/expr.h

```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/glfdc/bitvector.h>
+#include <torch/csrc/jit/codegen/cuda/glfdc/sexpr.h>
+
+#include <c10/util/Exception.h>
+
+#include <map>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+namespace glfdc {
+
+struct ExprDAG;
+
+// Expr - points subexpression and its owner expression DAG (ExprDAG)
+// NB: Expr should not outlive ExprDAG, since stores index to;
+// This is convinence class which wraps subexpression and also stores its DAG
+struct Expr {
+  const ExprDAG* dag_; // owner DAG
+  Operand subexpr_; // subexpression refered to inside
+};
+
+// Keeps DAG of all expressions reduction graphs
+struct ExprDAG {
+  // Maps unbound values (symbolic) cookies to index into unbound_values_
+  std::map<uintptr_t, std::size_t> unbound_lookup_;
+  // Stores all unbound values cookies
+  std::vector<uintptr_t> unbound_values_;
+
+  // List of subexpressions
+  std::vector<SExpr> subexpressions_;
+
+  // One bit reference count of subexpressions reuse for CSE
+  BitVector reused_subexpressions_;
+
+  // Number of reused subexpressions
+  std::size_t reuse_count_ = 0;
+
+ public:
+  // Adds single subexpression to DAG performing simple CSE
+  SExprRef addSubExpr(SExpr expr) {
+    size_t new_idx = subexpressions_.size();
+    subexpressions_.push_back(expr);
+    reused_subexpressions_.push_back(false);
+    TORCH_INTERNAL_ASSERT(
+        reused_subexpressions_.size() == subexpressions_.size());
+    return SExprRef{new_idx};
+  }
+
+  // Returns expression for subexpression reference
+  SExpr fetch(SExprRef e) const noexcept {
+    TORCH_INTERNAL_ASSERT(e.index_ < subexpressions_.size());
+    return subexpressions_[e.index_];
+  }
+
+  // Returns binding cookie for give symbolic value
+  uintptr_t getBinding(UnboundValue ubv) const noexcept // O(1)
+  {
+    TORCH_INTERNAL_ASSERT(ubv.index_ < unbound_values_.size());
+    return unbound_values_[ubv.index_];
+  }
+
+  // Returns number of reused subexpression inside DAG
+  std::size_t reuseCount() const noexcept {
+    return reuse_count_;
+  }
+
+  // Marks subexpression as reused for CSE
```

Comment:
What does it mean when a subexpression is reused?

Reply:
It was marked as common part of another expression by CSE. As a simple example, given expressions `(x+1)*7` and `(x+1)*y` the `(x+1)` subexpression is common part. Once evaluated once for given value of x the result is kept in glfdc::EvalState slot and doesn't need to be ever evaluated.

Reply:
Please leave it as a comment.